### ISSUE TITLE
AccessViolationException and app crash when working through proxy in multiple threads

### ISSUE THREAD

**Issue by kpdenis** [mozilla/geckodriver#1399]
## System

* Version: 0.23.0-win64
* Platform: Windows 7
* Firefox: 62.0.3 (64 bit)
* Selenium: 3.14.0

## Testcase
1. Create C# console app with code:

```
    class Program
    {
        static void Main(string[] args)
        {
            var proxies = new[]
            {
                "200.187.87.138:20183",
                "182.76.18.230:30073",
                "31.31.224.195:49160",
                "181.137.56.155:47756",
                "185.9.86.186:39345",
                "185.13.199.162:35702",
                "24.107.173.3:34370",
                "212.91.178.110:53281"
            };
            
            var threads = proxies.Select(x => new Thread(o => DoTest(x)))
                                 .ToList();

            foreach (var thread in threads)
                thread.Start();

            foreach (var thread in threads)
                thread.Join();
        }

        private static void DoTest(string proxy)
        {
            for (int i = 0; i < 10000; i++)
            {
                try
                {
                    using (var driver = CreateFireFox(proxy))
                    {
                        var r = Test(driver) ? "OK" : "FAIL";
                        Console.WriteLine($"{proxy} - [{r}]");
                    }
                }
                catch (Exception ex)
                {
                    Console.WriteLine(ex.Message);
                    // ignore Exception
                }
            }
        }

        private static bool Test(RemoteWebDriver driver)
        {
            try
            {
                // Do anything
                driver.Navigate().GoToUrl("https://www.seleniumhq.org/download/");
                return driver.FindElementsById("container").Any();
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                return false;
            }
        }

        private static RemoteWebDriver CreateFireFox(string proxy)
        {
            var execDir = Path.GetDirectoryName(Assembly.GetEntryAssembly().GetName().CodeBase.Replace("file:///", ""));
            if (execDir == null) throw new Exception("Exec dir not found");

            var firefoxDriverPath = Path.Combine(execDir, "Firefox");
            var service = FirefoxDriverService.CreateDefaultService(firefoxDriverPath);
            service.FirefoxBinaryPath = Path.Combine(firefoxDriverPath, "firefox.exe");

            var options = new FirefoxOptions
            {
                Profile = new FirefoxProfile()
            };

            // Manual, see https://developer.mozilla.org/en-US/docs/Mozilla/Preferences/Mozilla_networking_preferences
            options.Profile.SetPreference("network.proxy.type", 1);

            var data = proxy.Split(':');
            var httpProxy = data[0];
            var port = int.Parse(data[1]);

            options.Profile.SetPreference("network.proxy.http", httpProxy);
            options.Profile.SetPreference("network.proxy.http_port", port);
            options.Profile.SetPreference("network.proxy.ssl", httpProxy);
            options.Profile.SetPreference("network.proxy.ssl_port", port);

            return new FirefoxDriver(service, options, TimeSpan.FromMinutes(1));
        }
    }
```

2. copy FireFox v62.0.3 64 bit into bin/Debug/Firefox folder
3. copy geckodriver.exe (0.23.0-win64) into bin/Debug/Firefox folder
4. Run and wait

After working some time (about 5-10 minutes) this code raises exception "An unhandled exception of type 'System.AccessViolationException' occurred in Unknown Module." and app crash.

**Comment by andreastt** [comment]
`System.AccessViolationException` would stem from C#, but it could be that some syscall in geckodriver causes it. Without better reproduction steps I’m not sure there’s anything we can do here. I doubt a trace-level log would help, and it’s also not certain it is the result of a geckodriver syscall in the first place.

😔

**Comment by kpdenis** [comment]
Did you check described test case?
How long did it work without crash?

On my computer this code produce crash every run, just wait...

Exception message: "An unhandled exception of type 'System.AccessViolationException' occurred in Unknown Module.
Additional information: Attempted to read or write protected memory. This is often an indication that other memory is corrupt"

**Comment by whimboo** [comment]
Could you please attach a [trace-level log] from geckodriver? Read more about reporting actionable bugs in our [contribution guidelines].

If that is a crash of Firefox I would be kinda interested in getting this investigated. Also please try the same with the latest Firefox 63 beta which can be downloaded from https://www.mozilla.org/en-US/firefox/channel/desktop/#beta.

[trace-level log]: https://firefox-source-docs.mozilla.org/testing/geckodriver/geckodriver/TraceLogs.html
[contribution guidelines]: https://firefox-source-docs.mozilla.org/testing/geckodriver/geckodriver/Bugs.html

**Comment by kpdenis** [comment]
Here are logs.

Notice that log not finished on error, so the problem probably not at the end of log file.
Other threads proceeds write log before I stop the process.

[log.txt](https://github.com/mozilla/geckodriver/files/2496762/log.txt)
[log2.txt](https://github.com/mozilla/geckodriver/files/2496777/log2.txt)
[log3.txt](https://github.com/mozilla/geckodriver/files/2496784/log3.txt)

**Comment by whimboo** [comment]
What are the differences between those logs? Also how many instances of geckodriver are you starting with your test job? As it looks like a couple of them. How does it work when you only make use of a single geckodriver process?

**Comment by kpdenis** [comment]
This three logs are independent tests, all of them contains different crash moment.

There are 8 threads, each creates a new geckodriver and after making test, dispose it.
This works in cycle, geckodriver creates and disposes many times.

Constructor new FirefoxDriver(...) starts new geckodriver process, so there is max 8 geckodriver processes working in same moment.

**Comment by whimboo** [comment]
Well, as such the flow of log entries is intersected and hard to follow. Is there a way for you to also see this with a single thread only? Or does it only happen with multiple threads? If that is the case maybe you can isolate the the log output for each thread, and only upload one which shows the problem.
